axum/src/middleware/map_request.rs

```diff
@@ -0,0 +1,308 @@
+#![allow(missing_docs, unused_imports)]
+
+use crate::response::{IntoResponse, Response};
+use axum_core::extract::{FromRequest, FromRequestParts};
+use futures_util::future::BoxFuture;
+use http::Request;
+use std::{
+    any::type_name,
+    convert::Infallible,
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+};
+use tower::{util::BoxCloneService, ServiceBuilder};
+use tower_layer::Layer;
+use tower_service::Service;
+
+// TODO(david): naming
+pub trait IntoResult<T> {
+    fn into_result(self) -> Result<T, Response>;
+}
+
+impl<T, E> IntoResult<T> for Result<T, E>
+where
+    E: IntoResponse,
+{
+    fn into_result(self) -> Result<T, Response> {
+        self.map_err(IntoResponse::into_response)
+    }
+}
+
+impl<B> IntoResult<Request<B>> for Request<B> {
+    fn into_result(self) -> Result<Request<B>, Response> {
+        Ok(self)
+    }
+}
```

Comment:
This allows users to short-circuit, without requiring `try_*` versions. Though it might be easier to understand if we have the `try_*` versions.

Reply:
ahh, How would this cover things that are infallible?

Reply:
Just return `Request<B>` like https://github.com/tokio-rs/axum/pull/1408/files#diff-b565cc6c55fd226647bc08a856595f724ae48a0b5aef88774a11e5b6e36dbef6R265